Adds more scaffolding for experimental event API

trueadm opened the pull request:
This PR is a follow up to https://github.com/facebook/react/pull/15108. This adds the rest of the experimental event API scaffolding logic. Like the previous PR, this is only intended for internal FB testing right now. I also added flags around everything, so bundle size shouldn't regress other than a few bytes (if it does, I've done something wrong with DCE).

I've also added relevant fiber tests to ensure the warnings fire as expected.

sebmarkbage commented on packages/react-reconciler/src/ReactFiberBeginWork.js:
```diff
@@ -1923,6 +1926,70 @@ function updateContextConsumer(
+    let child = workInProgress.child;
+    while (child !== null) {
+      if (child.tag === HostText) {
+        warningWithoutStack(
```
This won’t warn if the node is nested in a function component for example. We could use the host context for this instead. We don’t have to do that in this PR but can you add a test for the warning and maybe a commented out test for the nested case? That way it’s easy to do that change later.

sebmarkbage replied:
Why exclude the stack here? Seems useful to have the stack in the warning.

sebmarkbage commented on packages/react-reconciler/src/ReactFiberBeginWork.js:
```diff
@@ -1923,6 +1926,70 @@ function updateContextConsumer(
+    while (child !== null) {
+      if (child.tag === HostText) {
+        childrenCount++;
+        invariant(
```
This should be a warning if it’s behind a DEV flag. Let’s also add a test that covers this.

Not only does it let us test the warning but also make sure the rest of the basic stuff here works too.

sebmarkbage commented on packages/react-reconciler/src/ReactFiberCompleteWork.js:
```diff
@@ -766,13 +769,17 @@ function completeWork(
     case EventTarget: {
-      markUpdate(workInProgress);
+      if (enableEventAPI) {
+        markUpdate(workInProgress);
```
This is fine as a start but longer term it would be good if we could avoid doing this in the commit phase for initial renders.

For host components we add events and stuff during render phase during initial mount (`current === null`). We only do stuff in the commit phase for updates.

trueadm replied:
In this case, the event target needs to mutate the direct DOM elements of the event target. This is only known till commit phase as the children might not have been created/attached in complete phase. Well at least that's the issue I ran into when testing this, maybe I'm missing something though.

trueadm commented:
I've addressed the feedback. I'll follow up on any other concerns in subsequent PRs. :)

trueadm commented on packages/react-dom/src/client/validateDOMNesting.js:
```diff
@@ -159,10 +162,18 @@ if (__DEV__) {
   };
 
-  updatedAncestorInfo = function(oldInfo, tag) {
+  updatedAncestorInfo = function(oldInfo, tag: string | Symbol) {
```
I originally did that, but it meant updating the host context object to include another property, which felt a bit over-engineered when technically event components are related to the DOM ancestor tree too.

trueadm replied:
Updated and changed it to use its own logic without using `updatedAncestorInfo`.

trueadm commented on packages/react-dom/src/shared/DOMNamespaces.js:
```diff
@@ -29,8 +33,11 @@ export function getIntrinsicNamespace(type: string): string {
-  type: string,
+  type: string | Symbol,
 ): string {
+  if (enableEventAPI && type === REACT_EVENT_COMPONENT_TYPE) {
```
My same comment applies. Splitting this function up might also make some parts more confusion. I'll give it another pass by though.

trueadm commented on packages/react-reconciler/src/ReactFiberBeginWork.js:
```diff
@@ -1923,6 +1926,63 @@ function updateContextConsumer(
+    'Event target components must be direct children of event components',
+  );
+  // These warnings only occur in DEV to reduce overhead in production
+  if (__DEV__ && nextProps.type === 'touch-hit') {
```
I can change it to a symbol. For now there is only one valid type and that is touch hit. I'll wrap the logic in a type.

gaearon replied:
I guess I got confused because in another place, what used to always be a string became `string | symbol`

trueadm replied:
Updated this and made it a symbol.

gaearon commented on packages/react-reconciler/src/ReactFiberHostContext.js:
```diff
@@ -79,7 +82,12 @@ function pushHostContext(fiber: Fiber): void {
   const context: HostContext = requiredContext(contextStackCursor.current);
-  const nextContext = getChildHostContext(context, fiber.type, rootInstance);
+  let type = fiber.type;
+  // For React event components, we get the typeof field (the symbol)
```
Why?

trueadm replied:
Because we then don't need to expose the fiber work tag logic in the renderer for the fiber.

sebmarkbage commented on packages/react-dom/src/client/ReactDOMHostConfig.js:
```diff
@@ -142,14 +148,17 @@ export function getRootHostContext(
   return namespace;
 }
 
 export function getChildHostContext(
```
It would be better to have a `getChildHostContextForEvent` function in the HostConfig. That way there can be a default noop version that just returns the parent context that will be inlined into beginWork.

We also don't have to consider the complexities of events when dealing with just simple host components.

sebmarkbage commented on packages/react-reconciler/src/ReactFiberHostContext.js:
```diff
@@ -79,7 +82,12 @@ function pushHostContext(fiber: Fiber): void {
+  const nextContext = getChildHostContext(context, type, rootInstance);
 
   // Don't push this Fiber's context unless it's unique.
   if (context === nextContext) {
```
If you create a separate pushHostContextForEvent function then for renderers where we call `getChildHostContextForEvent` and that's an identity function that just returns the same context, then this should be able to be DCE since it will know that the context is always the same.

sebmarkbage commented on packages/react-reconciler/src/ReactFiberBeginWork.js:
```diff
@@ -1923,6 +1929,64 @@ function updateContextConsumer(
+  );
+  const eventTargetType = nextProps.type;
+  // These warnings only occur in DEV to reduce overhead in production
+  if (__DEV__ && eventTargetType === REACT_EVENT_TARGET_TOUCH_HIT) {
```
This can't happen at all right now since nothing uses the Symbol. Can we get a test of how you expect this to be used even if it's a light version of some internal implementation details of the plugin?

trueadm commented on packages/react-reconciler/src/ReactFiberBeginWork.js:
```diff
@@ -1923,6 +1929,64 @@ function updateContextConsumer(
+  invariant(
+    parent !== null && parent.tag === Event,
+    'Event target components must be direct children of event components',
+  );
```
Your example would work, the parent needs to be an EventComponent and in the above case, it is.

trueadm commented:
@sebmarkbage I've added tests and moved all the logic checking into host context. This simplifies a bunch of cases and moves the checking logic into the renderer, which makes far more sense.

sebmarkbage commented on packages/react-dom/src/client/ReactDOMHostConfig.js:
```diff
@@ -165,6 +180,27 @@ export function getChildHostContext(
+          parentHostContext.$$typeof === REACT_EVENT_COMPONENT_TYPE,
+        'validateDOMNesting: React event targets must be direct children of event components.',
+      );
+      return {...parentHostContextDev, isEventTarget: true, hostNodeCount: 0};
```
Let's explicitly list out the props you want to copy here so it's clear what overrides are expected and what is expected to reset.

sebmarkbage commented on packages/react-dom/src/client/ReactDOMHostConfig.js:
```diff
@@ -165,6 +180,27 @@ export function getChildHostContext(
+): HostContext {
+  if (__DEV__) {
+    const parentHostContextDev = ((parentHostContext: any): HostContextDev);
+    if (event.$$typeof === REACT_EVENT_COMPONENT_TYPE) {
```
It might be nicer to pass a flag or the symbol as an argument and whatever else you need as an argument. That way the various host configs aren't so tied to the implementation details of the VDOM and the structure of this object. E.g. for createInstance we pass `type` and `props` as explicit arguments instead of passing the whole React Element.

sebmarkbage commented on packages/react-dom/src/client/ReactDOMHostConfig.js:
```diff
@@ -193,6 +229,17 @@ export function createInstance(
     const hostContextDev = ((hostContext: any): HostContextDev);
+    if (__DEV__ && enableEventAPI) {
+      if (hostContextDev.isEventTarget) {
+        (hostContextDev: any).hostNodeCount++;
```
Hm. This won't work. Mutation in here is not safe because we could do multiple passes over the parent or render them out of order in a Suspense case.

It might be better to place this warning in the place where you actually attach the event listeners and the error happens. I'm not sure where you're planning to attach the listeners but if you do it in the commit phase and plan on traversing the Fibers to find the DOM node, then you can do the warning there too. Another possible place is in the complete phase for the EventTarget, because that's when all children are guaranteed to have been completed.

trueadm commented:
@sebmarkbage I had to make the test `internal` because it writes to the feature flags to enable the events API. Tests that do this must always been internal it seems (according to comments in the config file for the test runner).

sebmarkbage commented on packages/react-dom/src/shared/DOMNamespaces.js:
```diff
@@ -31,6 +35,9 @@ export function getChildNamespace(
   parentNamespace: string | null,
   type: string,
 ): string {
+  if (enableEventAPI && type === REACT_EVENT_COMPONENT_TYPE) {
```
You can revert this whole block because it can't happen anymore. The `type` is a string and `REACT_EVENT_COMPONENT_TYPE` is a symbol.

sebmarkbage commented on packages/shared/getElementFromTouchHitTarget.js:
```diff
@@ -0,0 +1,62 @@
+
+export default function getElementFromTouchHitTarget(
+  targetFiber: Fiber,
+): null | Element {
```
The return type here should be `Instance` from HostConfig since it's whatever the host config defined it to be. Which happens to be Element for DOM but you're calling it from noop too.

NE-SmallTown commented on packages/react-reconciler/src/__tests__/ReactFiberEvents-test-internal.js:
```diff
@@ -0,0 +1,361 @@
+      expect(ReactNoop).toMatchRenderedOutput(<div>Hello world</div>);
+    });
+
+    it('should warn when an event component has a direct text child', () => {
```
has -> does't have?
